napari/layers/image/image.py

```diff
@@ -689,3 +689,34 @@ def contrast_limits(self, contrast_limits):
                 self.refresh()
             finally:
                 self._keep_auto_contrast = prev
+
+    def _calculate_value_from_ray(self, values):
+        values_within_clim = (values >= self.contrast_limits[0]) & (
+            values <= self.contrast_limits[1]
+        )
+        values_masked = np.where(values_within_clim, values, np.nan)
+        if np.all(np.isnan(values_masked)):
+            return None  # out of clims everywhere
+        if self.rendering == ImageRendering.TRANSLUCENT:
+            first_val = np.ravel(values_masked)[0]
+            return first_val if not np.isnan(first_val) else None
+        # these are probably not the same as how the gpu does it...
+        if self.rendering == ImageRendering.AVERAGE:
+            return np.nanmean(values_masked)
+        if self.rendering == ImageRendering.ADDITIVE:
+            # TODO: this is "broken" cause same pixel gets multisampled...
+            #       but it looks like it's also overdoing it in vispy vis too?
+            #       I don't know if there's a way to *not* do it...
+            return np.nansum(values_masked)
+        if self.rendering == ImageRendering.MIP:
+            return np.nanmax(values_masked)
+        if self.rendering == ImageRendering.MINIP:
+            return np.nanmin(values_masked)
+        if self.rendering == ImageRendering.ATTENUATED_MIP:
+            # TODO
```

Comment:
@jni this is the only one missing, but I never really wrapped my head around the attenuation code in vispy. Can you translate it to numpy for here?

Reply:
I should be able to, but I'm still not 100% we have it right — either here or in vispy — since the drops in brightness with decreasing quality came back. 😕 But actually implementing it in NumPy would help. 😊